libcxx/test/libcxx/time/time.zone/time.zone.db/rules.pass.cpp

```diff
@@ -0,0 +1,571 @@
+}
+
+static void test_invalid() {
+  test_exception("R", "corrupt tzdb: expected whitespace");
+
+  test_exception("R ", "corrupt tzdb: expected a string");
+
+  test_exception("R r", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r x", "corrupt tzdb: expected a digit");
+  test_exception("R r +", "corrupt tzdb: expected a digit");
+  test_exception("R r mx", "corrupt tzdb year: expected 'min' or 'max'");
+  test_exception("R r -32768", "corrupt tzdb year: year is less than the minimum");
+
+  test_exception("R r mix", "corrupt tzdb: expected whitespace");
+  test_exception("R r 0", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 x", "corrupt tzdb: expected a digit");
+  test_exception("R r 0 +", "corrupt tzdb: expected a digit");
+  test_exception("R r 0 mx", "corrupt tzdb year: expected 'min' or 'max'");
+
+  test_exception("R r 0 mix", "corrupt tzdb: expected whitespace");
+  test_exception("R r 0 1", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 1 X", "corrupt tzdb: expected character '-'");
+
+  test_exception("R r 0 1 -", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 1 - j", "corrupt tzdb month: invalid name");
+
+  test_exception("R r 0 1 - Ja", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 1 - Ja +", "corrupt tzdb weekday: invalid name");
+  test_exception("R r 0 1 - Ja 32", "corrupt tzdb day: value too large");
+  test_exception("R r 0 1 - Ja l", "corrupt tzdb: expected string 'last'");
+  test_exception("R r 0 1 - Ja last", "corrupt tzdb weekday: invalid name");
+  test_exception("R r 0 1 - Ja lastS", "corrupt tzdb weekday: invalid name");
+  test_exception("R r 0 1 - Ja S", "corrupt tzdb weekday: invalid name");
+  test_exception("R r 0 1 - Ja Su", "corrupt tzdb on: expected '>=' or '<='");
+  test_exception("R r 0 1 - Ja Su>", "corrupt tzdb: expected character '='");
+  test_exception("R r 0 1 - Ja Su<", "corrupt tzdb: expected character '='");
+  test_exception("R r 0 1 - Ja Su>=+", "corrupt tzdb: expected a non-zero digit");
+  test_exception("R r 0 1 - Ja Su>=0", "corrupt tzdb: expected a non-zero digit");
+  test_exception("R r 0 1 - Ja Su>=32", "corrupt tzdb day: value too large");
+
+  test_exception("R r 0 1 - Ja Su>=31", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 1 - Ja Su>=31 ", "corrupt tzdb: expected a digit");
+  test_exception("R r 0 1 - Ja Su>=31 +", "corrupt tzdb: expected a digit");
+
+  test_exception("R r 0 1 - Ja Su>=31 1", "corrupt tzdb: expected whitespace");
+  test_exception("R r 0 1 - Ja Su>=31 1a", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 1 - Ja Su>=31 1w 2", "corrupt tzdb: expected whitespace");
+  test_exception("R r 0 1 - Ja Su>=31 1w 2a", "corrupt tzdb: expected whitespace");
+
+  test_exception("R r 0 1 - Ja Su>=31 1w 2s", "corrupt tzdb: expected whitespace");
+  test_exception("R r 0 1 - Ja Su>=31 1w 2s ", "corrupt tzdb: expected a string");
+}
+
+#if 0
```

Comment:
What are we missing in the public API to be able to enable these tests via public APIs? It would be really nice (important even) to have these tests working in this patch.

Can't you `#include "include/tzdb/types.h"` here to test this? We do something similar for testing filesystem implementation details. It's not the cleanest way but IMO it's better than having `#if 0` for the time being.